packages/lib/src/interactions/utils.ts

```diff
@@ -65,3 +55,45 @@ export function getDefaultInteractions(
     SelectToZoom: { modifierKey: 'Control' },
   };
 }
+
+export function getEnclosedRectangle(startPoint: Vector2, endPoint: Vector2) {
+  // center = start + (end - start) / 2
+  const center = endPoint
+    .clone()
+    .sub(startPoint)
+    .divideScalar(2)
+    .add(startPoint);
+
+  return {
+    width: Math.abs(endPoint.x - startPoint.x),
+    height: Math.abs(endPoint.y - startPoint.y),
+    center,
+  };
+}
+
+export function clampRectangleToArea(
+  center: Vector2,
+  rectSize: Size,
+  areaSize: Size
+) {
+  const xBound = Math.max(areaSize.width / 2 - rectSize.width / 2, 0);
+  const yBound = Math.max(areaSize.height / 2 - rectSize.height / 2, 0);
+
+  return new Vector2(
+    clamp(center.x, -xBound, xBound),
+    clamp(center.y, -yBound, yBound)
+  );
+}
+
+export function clampRectangleToVis(
+  startPoint: Vector2,
+  endPoint: Vector2,
+  visSize: Size
+) {
+  const { center, ...rectSize } = getEnclosedRectangle(startPoint, endPoint);
+
+  const newCenter = clampRectangleToArea(center, rectSize, visSize);
+
+  const shift = newCenter.clone().sub(center);
+  return [startPoint, endPoint].map((v) => v.clone().add(shift));
+}
```

Comment:
Not really convinced on the function separation here.

I would like to keep `getEnclosedRectangle` to reuse it in `SelectToZoom` but  `clampRectangleToArea` can be inlined if we don't plan on reusing it in `moveCameraTo`.

Reply:
I'm curious as to how you would use `clampRectangleToArea` in `moveCameraTo`. The code is so similar, it'd be a shame to duplicate it. But if you think it's not worth it, then inlining would indeed make sense here.

Reply:
My first implementation felt too convoluted which why I did not include it in the PR.

I tried again this morning and it went smoothly. I guess that's why I should not try too hard Friday afternoon :joy: